### PR TITLE
Add retries to ModulesAreDisposedWhenProcessIsDisposed

### LINKED ISSUE #68035: System.Diagnostics.Tests.ProcessModuleTests.ModulesAreDisposedWhenProcessIsDisposed test is failing
Build: https://dnceng.visualstudio.com/public/_build/results?buildId=1711437&view=ms.vss-test-web.build-test-results-tab&runId=46535502&resultId=129201&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

Configuration: `net7.0-windows-Release-x86-CoreCLR_release-Windows.7.Amd64.Open`

```
  Discovering: System.Diagnostics.Process.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Diagnostics.Process.Tests (found 258 of 281 test cases)
  Starting:    System.Diagnostics.Process.Tests (parallel test collections = on, max threads = 2)
    System.Diagnostics.Tests.ProcessModuleTests.ModulesAreDisposedWhenProcessIsDisposed [FAIL]
      System.ComponentModel.Win32Exception : Only part of a ReadProcessMemory or WriteProcessMemory request was completed.
      Stack Trace:
        /_/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs(239,0): at System.Diagnostics.NtProcessManager.EnumProcessModulesUntilSuccess(SafeProcessHandle processHandle, IntPtr[] modules, Int32 size, Int32& needed)
        /_/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs(111,0): at System.Diagnostics.NtProcessManager.GetModules(Int32 processId, Boolean firstModuleOnly)
        /_/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Windows.cs(380,0): at System.Diagnostics.NtProcessManager.GetModules(Int32 processId)
        /_/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Windows.cs(165,0): at System.Diagnostics.ProcessManager.GetModules(Int32 processId)
        /_/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs(315,0): at System.Diagnostics.Process.get_Modules()
        /_/src/libraries/System.Diagnostics.Process/tests/ProcessModuleTests.cs(79,0): at System.Diagnostics.Tests.ProcessModuleTests.ModulesAreDisposedWhenProcessIsDisposed()
           at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
        /_/src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.CoreCLR.cs(379,0): at System.Reflection.RuntimeMethodInfo.InvokeNonEmitUnsafe(Object obj, IntPtr* arguments, Span`1 argsForTemporaryMonoSupport, BindingFlags invokeAttr)
    System.Diagnostics.Tests.ProcessStartInfoTests.ShellExecute_Nano_Fails_Start [SKIP]
      Condition(s) not met: "IsWindowsNanoServer"
Invalid number of parameters
0 File(s) copied
  Finished:    System.Diagnostics.Process.Tests
=== TEST EXECUTION SUMMARY ===
   System.Diagnostics.Process.Tests  Total: 318, Errors: 0, Failed: 1, Skipped: 1, Time: 33.721s
```

### PR DESCRIPTION
**Pull request by danmoseley**
Fix https://github.com/dotnet/runtime/issues/68035

It seems that EnumProcessModules is prone to returning ERROR_PARTIAL_COPY when invoked on a process that is in the process of starting (in some terms). We already have 50ms of retries inside the product for this; possibly that should be more retries, and only for ERROR_PARTIAL_COPY. However I'm reluctant to to modify that without more evidence. Meantime, just adding retries to this one test that calls it immediately after starting a process.

### CODE REVIEW

**Comment by danmoseley** [comment]
Failure is the indomitable https://github.com/dotnet/runtime/issues/67349

File: src/libraries/System.Diagnostics.Process/tests/ProcessModuleTests.cs [hunk 1]
```diff
@@ -76,7 +77,15 @@ public void ModulesAreDisposedWhenProcessIsDisposed()
...
 
-            ProcessModuleCollection modulesCollection = process.Modules;
+            // Very rarely, this call will fail with ERROR_PARTIAL_COPY; it only happened
+            // so far on this particular test, the only one that creates a new process.
+            // Assumption is that we need to give a little extra time.
+            ProcessModuleCollection modulesCollection = null;
+            RetryHelper.Execute(() =>
+            {
+                modulesCollection = process.Modules;
+            }, maxAttempts:5, backoffFunc:null, retryWhen:e => { return (e.GetType() == typeof(Win32Exception)); });
```
Comment (adamsitnik) [inline src/libraries/System.Diagnostics.Process/tests/ProcessModuleTests.cs:87]: minor (and opinionated nit):

```suggestion
            }, maxAttempts: 5, backoffFunc: null, retryWhen: e => e.GetType() == typeof(Win32Exception));
```

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @danmoseley !

File: src/libraries/System.Diagnostics.Process/tests/ProcessModuleTests.cs (hunk 1 above)
Comment (danmoseley) [reply src/libraries/System.Diagnostics.Process/tests/ProcessModuleTests.cs:87]: 51 matches for `, \w+:\w+,` and 2426 for `, \w+: \w+,` -- you win :) @sharwell is there a rule we can enable to force a space in these?

**Comment by danmoseley** [comment]
failures are https://github.com/dotnet/runtime/issues/68293

**Pull request merged**